### PR TITLE
New: Use process @abstract when processing @return (fixes #5941)

### LINKED ISSUE #5941: Ignore invalid return when jsdoc @abstract is present
**What version of ESLint are you using?**
2.8.0

**What parser (default, Babel-ESLint, etc.) are you using?**
Default

**What did you do? Please include the actual source code causing the issue.**

I have somewhat of an abstract class where I have a method such as this:

```javascript
    /**
     * Creates a new OAuth grant using the provided query and permissions. At
     * this point the query and permissions will already have been validated.
     * @abstract
     * @param  {Object} query
     * @param  {String[]} permissions
     * @return {Promise.<{code: String, state: String}>}
     */
    createOAuth (query, permissions) {
        throw new Error('not implemented');
    }
```

**What did you expect to happen?**

Ideally with JSDoc enabled, eslint should ignore invalid returns when `@abstract` is set.

**What actually happened? Please include the actual, raw output from ESLint.**

ESLint complains about a the missing return in the function body eg. `Unexpected @return tag`.

### PR DESCRIPTION
**Pull request by SimonSchick**
Allow @return(s) without return keyword in function when @abstract/virtual is present

My first ESLint PR, please be gentle 😛

### CODE REVIEW

**Comment by platinumazure** [comment]
Hi @SimonSchick, thanks for the pull request and congrats on your first contribution!

One thing for next time: We generally use "New:" prefix in commit message only for completely new functionality; new facets in existing functionality (such as new rule options) generally get an "Update:" prefix. You can use the issue labels to figure out the best commit summary prefix.

You can peruse our [Working on Issues](http://eslint.org/docs/developer-guide/contributing/working-on-issues) document for more information. Thanks again for your contribution!

**Comment by SimonSchick** [comment]
Ah yes, makes sense.
Will I have to change my commit in order for it to be merged?

**Comment by platinumazure** [comment]
Until recently, the answer would have been yes, but with the new GitHub squash-merge functionality, you might not have to. :smile: Let's see what someone who is more in the know will say about that.

**Comment by BYK** [comment]
LGTM

File: docs/rules/valid-jsdoc.md [hunk 1]
```diff
@@ -26,6 +26,7 @@ This rule aims to prevent invalid and incomplete JSDoc comments. It will warn wh
 * A comment for a function is missing `@returns`
 * A parameter has no associated `@param` in the JSDoc comment
 * `@param`s are out of order with named arguments
+* `@return` is used on a function with no `return` keyword without `@abstract` or `@virtual`
```
Comment (nzakas) [inline docs/rules/valid-jsdoc.md]: I think we can just remove this. This lost already doesn't mention everything the rule does, and this bullet is a bit hard to understand.

File: tests/lib/rules/valid-jsdoc.js [hunk 2]
```diff
@@ -470,6 +470,28 @@ ruleTester.run("valid-jsdoc", rule, {
             parserOptions: {
                 ecmaVersion: 6
             }
+        },
+
+        // abstract
+        {
```
Comment (nzakas) [inline tests/lib/rules/valid-jsdoc.js]: Can you add some tests with `requireReturn` set to true?

**Comment by nzakas** [comment]
@SimonSchick overall looks good, just a couple of small things to take care of.

File: tests/lib/rules/valid-jsdoc.js (hunk 2 above)
Comment (SimonSchick) [inline tests/lib/rules/valid-jsdoc.js]: Not exactly sure what you want to see here, but I added 2 more tests.

**Comment by SimonSchick** [comment]
I updated the PR.

**Comment by nzakas** [comment]
Lgtm

@SimonSchick thanks, and congrats on your first PR! 
despicable-me-minions4 gif4

**Comment by SimonSchick** [comment]
Can you give me any information when the next release will happen as a thanks? 😄

**Comment by nzakas** [comment]
Our releases are every other Friday. The next one is this coming Friday, April 29.

**Comment by SimonSchick** [comment]
Ty @nzakas

**Pull request merged**